src/analysis/stack-lattice.h

```diff
@@ -0,0 +1,187 @@
+#ifndef wasm_analysis_stack_lattice_h
+#define wasm_analysis_stack_lattice_h
+
+#include <deque>
+#include <optional>
+
+#include "lattice.h"
+
+namespace wasm::analysis {
+
+// Models the WASM value stack
+// A lattice for modelling stack-related phenomena. It is templated
+// on a lattice representing the contents of each stack element.
+template<typename StackElementLattice> class StackLattice {
+  static_assert(is_lattice<StackElementLattice>);
+  StackElementLattice& stackElementLattice;
+
+public:
+  StackLattice(StackElementLattice& stackElementLattice)
+    : stackElementLattice(stackElementLattice) {}
+
+  class Element {
+    // The top lattice is
+    std::optional<std::deque<typename StackElementLattice::Element>>
+      stackValue = std::deque<typename StackElementLattice::Element>();
+
+    // Indicates if the lattice element is the top element, which theoretically
+    // would be an infinite-height stack. In practice, if this flag is raised,
+    // we ignore the actual stack and treat it like the top element.
+
+  public:
+    bool isTop() const { return !stackValue.has_value(); }
+    bool isBottom() const {
+      return stackValue.has_value() && stackValue->empty();
+    }
+    void setToTop() { stackValue.reset(); }
+
+    typename StackElementLattice::Element& stackTop() {
+      return stackValue->back();
+    }
+
+    void push(typename StackElementLattice::Element&& element) {
+      if (stackValue.has_value() &&
+          (!stackValue->empty() || !element.isBottom())) {
+        stackValue->push_back(std::move(element));
+      }
+    }
+
+    void push(const typename StackElementLattice::Element& element) {
+      if (stackValue.has_value() &&
+          (!stackValue->empty() || !element.isBottom())) {
+        stackValue->push_back(std::move(element));
+      }
+    }
+
+    typename StackElementLattice::Element pop() {
+      typename StackElementLattice::Element value = stackValue->back();
+      stackValue->pop_back();
+      return value;
+    }
+
+    // When taking the LUB, we take the LUBs of the elements of the suffixes of
+    // each stack. This is because when we have stack1 = [a] and stack2 = [b',
+    // a'] which is from a scope which encloses the scope of stack1, we want the
+    // result to be [b', LUB(a, a')]. We can alternatively imagine that each
```

Comment:
Why do we want that? I feel there is another step between this and using it somewhere that I don't see explained yet. Maybe an example could help?

Also, what would we want to happen to these two stacks? `[b, a] ; [b', a']`

Reply:
Added some more clarification, here, and at the top of the file. For the case you gave,
`LUB([b, a], [b', a']) = [LUB(b,b'), LUB(a,a')]`